sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/CreateHiveTableAsSelectCommand.scala

```diff
@@ -95,9 +77,116 @@ case class CreateHiveTableAsSelectCommand(
     Seq.empty[Row]
   }
 
+  // Returns `DataWritingCommand` used to write data when the table exists.
+  def writingCommandForExistingTable(
+    catalog: SessionCatalog,
+    tableDesc: CatalogTable): DataWritingCommand
+
+  // Returns `DataWritingCommand` used to write data when the table doesn't exist.
+  def writingCommandForNewTable(
+    catalog: SessionCatalog,
+    tableDesc: CatalogTable): DataWritingCommand
+
   override def argString: String = {
     s"[Database:${tableDesc.database}, " +
     s"TableName: ${tableDesc.identifier.table}, " +
     s"InsertIntoHiveTable]"
   }
 }
+
+/**
+ * Create table and insert the query result into it.
+ *
+ * @param tableDesc the table description, which may contain serde, storage handler etc.
+ * @param query the query whose result will be insert into the new relation
+ * @param mode SaveMode
+ */
+case class CreateHiveTableAsSelectCommand(
+    tableDesc: CatalogTable,
+    query: LogicalPlan,
+    outputColumnNames: Seq[String],
+    mode: SaveMode)
+  extends CreateHiveTableAsSelectBase {
+
+  override def writingCommandForExistingTable(
+      catalog: SessionCatalog,
+      tableDesc: CatalogTable): DataWritingCommand = {
+    InsertIntoHiveTable(
+      tableDesc,
+      Map.empty,
```

Comment:
Why this is empty? Do we have a test case for partitioning tables?

Reply:
Could you point it out? I want to ensure it is covered

Reply:
https://github.com/apache/spark/blob/8534d753ecb21ea64ffbaefb5eaca38ba0464c6d/sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala#L686-L697

Reply:
Does this work?
```
    withTable("hive_test") {
      withSQLConf(
          "hive.exec.dynamic.partition.mode" -> "nonstrict") {
        val df = Seq(("a", 100)).toDF("part", "id")
        df.write.format("hive").partitionBy("part")
          .mode("overwrite").saveAsTable("hive_test")
        df.write.format("hive").partitionBy("part")
          .mode("append").saveAsTable("hive_test")
      }
    }
```

Reply:
Yes, I've discussed with @cloud-fan. DataFrameWriter allows dynamic partition on Hive CTAS. For now seems this syntax is not allowed by SparkSqlParser.

Reply:
or open a new PR to allow CTAS for partitioned hive table first?

Reply:
That is different issue. I think it is better to have a separate PR to address it.

Reply:
Please hold this PR. We need to fix the regression first.

Reply:
https://github.com/apache/spark/pull/23255